File: app/lib/affine.dart
import 'package:vector_math/vector_math.dart';

class AffineTransform {
  final Matrix2 _matrix;
  final Vector2 _constant;

  AffineTransform(Matrix2 matrix, Vector2 constant)
      : _matrix = matrix,
        _constant = constant;

  Vector2 apply(Vector2 vector) {
    return _matrix * vector + _constant;
  }

  AffineTransform compose(AffineTransform other) {
    return AffineTransform(
        _matrix * other._matrix, _matrix * other._constant + _constant);
  }

  AffineTransform then(AffineTransform other) {
    return other.compose(this);
  }

  AffineTransform inverse() {
    var inverse = Matrix2.zero()..copyInverse(_matrix);
    return AffineTransform(inverse, -inverse * _constant);
  }

  factory AffineTransform.identity() {
    return AffineTransform(Matrix2.identity(), Vector2.zero());
  }
}

File: app/src/math/affine.ts
import { Matrix2 } from "./matrix2";
import { Vector2 } from "./vector2";

export class AffineTransform {
  public constructor(
    public readonly matrix: Matrix2,
    public readonly constant: Vector2
  ) {}

  public static identity(): AffineTransform {
    return new AffineTransform(Matrix2.identity(), Vector2.zero());
  }

  public apply(vector: Vector2): Vector2 {
    return this.matrix.multiplyVector(vector).add(this.constant);
  }

  public compose(other: AffineTransform): AffineTransform {
    return new AffineTransform(
      this.matrix.multiply(other.matrix),
      this.apply(other.constant)
    );
  }

  public then(other: AffineTransform): AffineTransform {
    return other.compose(this);
  }

  public inverse(): AffineTransform {
    const inverse = this.matrix.inverse();
    return new AffineTransform(
      inverse,
      inverse.scale(-1).multiplyVector(this.constant)
    );
  }
}
